sysv-pdp11_tape/man/u_man/man1/cut.1
.if t .ds ' \h@.05m@\s+4\v@.333m@\'\v@-.333m@\s-4\h@.05m@
.if n .ds ' '
.if t .ds ` \h@.05m@\s+4\v@.333m@\`\v@-.333m@\s-4\h@.05m@
.if n .ds ` `
.TH CUT 1
.SH NAME
cut \- cut out selected fields of each line of a file
.SH SYNOPSIS
\fBcut \-c\fP\^list [\|file1 file2 ...]
.br
\fBcut \-f\fP\^list [\fB\-d\fP\|char\|] [\fB\-s\fP] [\|file1 file2 ...]
.SH DESCRIPTION
Use
.I cut\^
to cut out columns from a table or 
fields from each line of a file; in data base parlance, it implements
the projection of a relation.
The fields as specified by
.I list\^
can be fixed length,
i.e., character positions as on a punched card
(\fB\-c\fP option), or the length can vary from line to line
and be marked with a field delimiter character like
.I tab\^
(\fB\-f\fP option).
.I Cut\^
can be used as a filter;
if no files are given, the standard input is used.
.PP
The meanings of the options are:
.br
.TP "\w'\-d\ char\ \ 'u"
.I list\^
A comma-separated 
list of integer field numbers (in increasing order),
with optional \fB\-\fP to indicate ranges
as in the 
.B \-o
option of
.IR nroff / troff
for page ranges;
e.g.,
.BR 1,4,7 ;
.BR 1\-3,8 ;
.B \-5,10
(short for \fB1\-5,10\fP); or
.B 3\-
(short for third through last field).
.TP
\fB\-c\fP\^\fIlist\fP
The
.I list\^
following 
.B \-c
(no space) specifies character
positions (e.g.,
.B \-c1\-72
would pass the first 72 characters
of each line).
.TP
\fB\-f\fP\^\fIlist\fP
The \fIlist\fP following 
.B \-f
is a list of fields
assumed to be separated in the file by a delimiter character (see 
.B \-d
);
e.g. ,
.B \-f1,7
copies the first and seventh field only.
Lines with no field delimiters will be passed through intact (useful
for table subheadings), unless 
.B \-s
is specified.
.TP
\fB\-d\fP\^\fIchar\fP
The character following 
.B \-d
is the field delimiter
(\c
.B \-f
option only).
Default is
.IR tab .
Space or other characters with special meaning to the shell must be quoted.
.TP
.B \-s
Suppresses lines with no delimiter characters in case of
.B \-f
option.
Unless specified, lines with no delimiters will be passed through untouched.
.PP
Either the
.B \-c
or 
.B \-f
option must be specified.
.SH HINTS
Use
.IR grep (1)
to make horizontal ``cuts''
(by context) through a file, or
.IR paste (1)
to put files together column-wise
(i.e., horizontally).
To reorder columns in a table, use
.I cut\^
and
.IR paste .
.SH EXAMPLES
.TP 2.25i
cut \-d: \-f1,5 /etc/passwd
mapping of user \s-1ID\s0s to names
.TP
name=\*`who am i | cut \-f1 \-d" "\*`
to set 
.B name
to current login
name.
.SH DIAGNOSTICS
.PP
.TP 1.25i
.I "line too long\^"
A line can have no more than 511 characters or fields.
.TP
.I "bad list for c\|/\|f option\^"
Missing 
.B \-c
or 
.B \-f
option or incorrectly specified
.IR list .
No error occurs if a line has fewer fields than the
.I list\^
calls for.
.TP
.I "no fields\^"
The
.I list\^
is empty.
.SH SEE ALSO
grep(1),
paste(1).
.\"	@(#)cut.1	5.2 of 5/18/82


sysvr3/31/usr/src/cmd/cut/cut.c
/*	Copyright (c) 1984 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)cut:cut.c	1.11"
#
/* cut : cut and paste columns of a table (projection of a relation) */
/* Release 1.5; handles single backspaces as produced by nroff    */
# include <stdio.h>	/* make: cc cut.c */
# include <ctype.h>
# define NFIELDS 1024	/* max no of fields or resulting line length */
# define BACKSPACE '\b'

int strcmp(), atoi();
void exit();


main(argc, argv)
int	argc;
char	**argv;
{
	extern int	optind;
	extern char	*optarg;
	static char	usage[] = "Usage: cut [-s] [-d<char>] {-c<list> | -f<list>} file ...";
	static char	cflist[] = "bad list for c/f option";
	register int	c;
	register char	*p1, *rbuf;
	register char	*p, *list;
	int	del = '\t';
	int	num, j, count, poscnt, r, s;
	int	endflag, supflag, cflag, fflag, backflag, filenr;
	static  int	sel[NFIELDS];
	char 	buf[NFIELDS];
	char	*p2, outbuf[NFIELDS];
	FILE	*inptr;
	supflag = cflag = fflag = r = num = s = 0;


	while((c = getopt(argc, argv, "c:d:f:s")) != EOF)
		switch(c) {
			case 'c':
				if (fflag)
					diag(usage);
				cflag++;
				list = optarg;
				break;
			case 'd':
				if (strlen(optarg) > 1)
					diag("no delimiter");
				else
					del = (int)*optarg;
				break;
			case 'f':
				if (cflag)
					diag(usage);
				fflag++;
				list = optarg;
				break;
			case 's':
				supflag++;
				break;
			case '?':
				diag(usage);
		}

	argv = &argv[optind];
	argc -= optind;

	if (!(cflag || fflag))
		diag(cflist);

	do {
		p = list;
		switch(*p) {
			case '-':
				if (r)
					diag(cflist);
				r = 1;
				if (num == 0)
					s = 1;
				else {
					s = num;
					num = 0;
				}
				break;
			case '\0' :
			case ','  :
				if (num >= NFIELDS)
					diag(cflist);
				if (r) {
					if (num == 0)
						num = NFIELDS - 1;
					if (num < s)
						diag(cflist);
					for (j = s; j <= num; j++)
						sel[j] = 1;
				} else
					sel[num] = (num > 0 ? 1 : 0);
				s = num = r = 0;
				if (*p == '\0')
					continue;
				break;
			default:
				if (!isdigit(*p))
					diag(cflist);
				num = atoi(p);
				while (isdigit(*list))
					list++;
				continue;
		}
		list++;
	}while (*p != '\0');
	for (j=0; j < NFIELDS && !sel[j]; j++);
	if (j >= NFIELDS)
		diag("no fields");

	filenr = 0;
	do {	/* for all input files */
		if ( argc == 0 || strcmp(argv[filenr],"-") == 0 )
			inptr = stdin;
		else
			if ((inptr = fopen(argv[filenr], "r")) == NULL) {
				fprintf(stderr, "cut: WARNING: cannot open %s\n", argv[filenr]);
				continue;
			}
		endflag = 0;
		do {	/* for all lines of a file */
			count = poscnt = backflag = 0;
			p1 = &outbuf[0] - 1 ;
			p2 = p1;
			rbuf = buf;
			if ((fgets(buf, NFIELDS, inptr)) == NULL) {
				endflag = 1;
				continue;
			}
			do { 	/* for all char of the line */
				if (rbuf >= &buf[NFIELDS])
					diag("line too long");
				if (*rbuf != '\n')
					*++p1 = *rbuf;
				if (cflag && (*rbuf == BACKSPACE))
					backflag++;
				else if (!backflag)
					poscnt += 1;
				else
					backflag--;
				if ( backflag > 1 )
					diag("cannot handle multiple adjacent backspaces\n");
				if (*rbuf == '\n' && count > 0  || *rbuf == del || cflag) {
					count += 1;
					if (fflag)
						poscnt = count;
					if (sel[poscnt])
						p2 = p1;
					else
						p1 = p2;
				}
			} while (*rbuf++ != '\n');
			if ( !endflag && (count > 0 || !supflag)) {
				if (*p1 == del && !sel[count])
					*p1 = '\0'; /*suppress trailing delimiter*/
				else
					*++p1 = '\0';
				puts(outbuf);
			}
		} while (!endflag);
		fclose(inptr);
	} while (++filenr < argc);
	exit(0);
}


diag(s)
char	*s;
{
	fprintf(stderr, "cut: ERROR: %s\n", s);
	exit(2);
}
